### ISSUE TITLE
Non-terminating loop in ScriptableObject.getTopLevelScope()

### ISSUE THREAD

**Issue by hrj** [apigee/rhino#23]
Hi,

Thanks for maintaining this fork of Rhino. We are using this fork in [our project](https://github.com/UprootLabs/gngr). Everything mostly works fine except this issue.

On certain JS heavy websites like Reddit, the JS engine goes into an infinite loop in this function:
```java
public static Scriptable getTopLevelScope(Scriptable obj)
{
  for (;;) {
    Scriptable parent = obj.getParentScope();
    if (parent == null) {
      return obj;
    }
    obj = parent;
  }
}
```

During the infinite loop, the stack looks like this:
```
Daemon Thread [JS Scheduler] (Suspended)	
	ScriptableObject.getTopLevelScope(Scriptable) line: 2125	
	ScriptableObject.getObjectPrototype(Scriptable) line: 2055	
	Arguments.<init>(NativeCall) line: 29	
	NativeCall.<init>(NativeFunction, Scriptable, Object[]) line: 55	
	ScriptRuntime.createFunctionActivation(NativeFunction, Scriptable, Object[]) line: 3196	
	Interpreter.initFrame(Context, Scriptable, Scriptable, Object[], double[], int, int, InterpretedFunction, Interpreter$CallFrame, Interpreter$CallFrame) line: 2745	
	Interpreter.interpretLoop(Context, Interpreter$CallFrame, Object) line: 1398	
	Interpreter.interpret(InterpretedFunction, Context, Scriptable, Scriptable, Object[]) line: 815	
	InterpretedFunction.call(Context, Scriptable, Scriptable, Object[]) line: 109	
	BoundFunction.call(Context, Scriptable, Scriptable, Object[]) line: 53	
	ScriptRuntime.applyOrCall(boolean, Context, Scriptable, Scriptable, Object[]) line: 2441	
	BaseFunction.execIdCall(IdFunctionObject, Context, Scriptable, Scriptable, Object[]) line: 287	
	IdFunctionObject.call(Context, Scriptable, Scriptable, Object[]) line: 97	
	Interpreter.interpretLoop(Context, Interpreter$CallFrame, Object) line: 1473	
	Interpreter.interpret(InterpretedFunction, Context, Scriptable, Scriptable, Object[]) line: 815	
	InterpretedFunction.call(Context, Scriptable, Scriptable, Object[]) line: 109	
	SecurityControllerImpl.lambda$0(Callable, Context, Scriptable, Scriptable, Object[]) line: 56	
	1601952727.run() line: not available	
	AccessController.doPrivileged(PrivilegedAction<T>, AccessControlContext) line: not available [native method]	
	SecurityControllerImpl.callWithDomain(Object, Context, Callable, Scriptable, Scriptable, Object[]) line: 59	
	Interpreter.interpret(InterpretedFunction, Context, Scriptable, Scriptable, Object[]) line: 802	
	InterpretedFunction.call(Context, Scriptable, Scriptable, Object[]) line: 109	
	Window$MyContextFactory(ContextFactory).doTopCall(Callable, Context, Scriptable, Scriptable, Object[]) line: 394	
	ScriptRuntime.doTopCall(Callable, Context, Scriptable, Scriptable, Object[]) line: 3102	
	InterpretedFunction.call(Context, Scriptable, Scriptable, Object[]) line: 107	
	XMLHttpRequest.lambda$2(Function) line: 231	
	273358880.run() line: not available	
	Window$JSRunnableTask.run() line: 253	
	Window$JSScheduler$1.run() line: 333	
	AccessController.doPrivileged(PrivilegedAction<T>, AccessControlContext) line: not available [native method]	
	Window$JSScheduler.run() line: 339	
```

We haven't enabled optimization.

This problem doesn't arise if we setup the rhino supplied debugger or even our own debugger that returns dummy debug frames.

Do you have any clue as to what could be the cause? Could this be a race-condition? Can the problem be on our side of the code (improper synchronisation, etc).

I realize it might be hard to tell without sufficient context; I am just trying a lucky shot.

Please let me know if you happen to have some clues.

**Comment by eshepelyuk** [comment]
I assume it could be caused by the fact that some of JS objects returns non-null for `getParentScope`.
Do you have possibility to gather such debug information ?

**Comment by hrj** [comment]
Yes, I am trying to get more debug information. It's a little tricky since the bug is not easily reproducible. Will update here soon.

**Comment by hrj** [comment]
One thing I can see is, there is a circular chain of parents. There are two objects which are parents of each other. This results in the infinite loop.

However, I am not sure what causes the circular chain. Still debugging.

**Comment by hrj** [comment]
I was able to trace the problem to a suspect in our code. In a nut shell, the top level Window scope's parent scope gets set to a NativeFunction object, whose parent-scope in turn is the Window. This creates circular chain. Preventing this from happening avoids the problem; though I am yet to wrap my head fully around it.

Sorry for the noise here. I was working on this problem on and off for the last couple of months, before I posted it here. And by fluke, I seem to have found the culprit just after posting.
